File: Wikipedia/Code/WMFAuthManagerInfoFetcher.m
#import "WMFAuthManagerInfoFetcher.h"
#import "AFHTTPSessionManager+WMFDesktopRetry.h"
#import "AFHTTPSessionManager+WMFConfig.h"
#import "MWNetworkActivityIndicatorManager.h"
#import "WMFMantleJSONResponseSerializer.h"
#import "WMFAuthManagerInfo.h"

NS_ASSUME_NONNULL_BEGIN

@interface WMFAuthManagerInfoFetcher ()
@property(nonatomic, strong) AFHTTPSessionManager *operationManager;
@end

@implementation WMFAuthManagerInfoFetcher

- (instancetype)init {
  self = [super init];
  if (self) {
    AFHTTPSessionManager *manager = [AFHTTPSessionManager wmf_createDefaultManager];
    manager.responseSerializer = [WMFMantleJSONResponseSerializer serializerForInstancesOf:[WMFAuthManagerInfo class] fromKeypath:@"query"];
    self.operationManager = manager;
  }
  return self;
}

- (BOOL)isFetching {
  return [[self.operationManager operationQueue] operationCount] > 0;
}

- (void)fetchAuthManagerCreationAvailableForSiteURL:(NSURL *)siteURL success:(WMFAuthManagerInfoBlock)success failure:(WMFErrorHandler)failure {
  [self fetchAuthManagerAvailableForSiteURL:siteURL type:@"create" success:success failure:failure];
}

- (void)fetchAuthManagerLoginAvailableForSiteURL:(NSURL *)siteURL success:(WMFAuthManagerInfoBlock)success failure:(WMFErrorHandler)failure {
  [self fetchAuthManagerAvailableForSiteURL:siteURL type:@"login" success:success failure:failure];
}

- (void)fetchAuthManagerAvailableForSiteURL:(NSURL *)siteURL type:(NSString *)type success:(WMFAuthManagerInfoBlock)success failure:(WMFErrorHandler)failure {
  NSDictionary *params = @{
    @"action" : @"query",
    @"meta" : @"authmanagerinfo",
    @"format" : @"json",
    @"amirequestsfor" : type
  };

  [self.operationManager wmf_GETAndRetryWithURL:siteURL
      parameters:params
      retry:NULL
      success:^(NSURLSessionDataTask *operation, id responseObject) {
        [[MWNetworkActivityIndicatorManager sharedManager] pop];
        success(responseObject);
      }
      failure:^(NSURLSessionDataTask *operation, NSError *error) {
        [[MWNetworkActivityIndicatorManager sharedManager] pop];
        failure(error);
      }];
}

@end

NS_ASSUME_NONNULL_END

File: Wikipedia/Code/WMFAuthManagerInfoFetcher.swift

import Foundation

extension WMFAuthManagerInfo2 {
    func captchaId() -> String? {
        return requests?.first(where:{$0.id! == "CaptchaAuthenticationRequest"})?.fields?.captchaId?.value
        
    }
    func captchaInfo() -> String? {
        return requests?.first(where:{$0.id! == "CaptchaAuthenticationRequest"})?.fields?.captchaInfo?.value
    }
}

class WMFAuthManagerInfoFetcher2 {
    private let manager = AFHTTPSessionManager.wmf_createDefault()
    
    func isFetching() -> Bool {
        return manager!.operationQueue.operationCount > 0
    }
    
    func fetchAuthManagerCreationAvailableForSiteURL(_ siteURL: URL, completion: WMFURLSessionDataTaskSuccessHandler, failure: WMFURLSessionDataTaskFailureHandler){
        fetchAuthManagerAvailableForSiteURL(siteURL, type: "create", completion: completion, failure: failure)
    }

    func fetchAuthManagerLoginAvailableForSiteURL(_ siteURL: URL, completion: WMFURLSessionDataTaskSuccessHandler, failure: WMFURLSessionDataTaskFailureHandler){
        fetchAuthManagerAvailableForSiteURL(siteURL, type: "login", completion: completion, failure: failure)
    }
    
    private func fetchAuthManagerAvailableForSiteURL(_ siteURL: URL, type: String, completion: WMFURLSessionDataTaskSuccessHandler, failure: WMFURLSessionDataTaskFailureHandler){
        let manager = AFHTTPSessionManager(baseURL: siteURL)
        
        manager.responseSerializer = WMFMantleJSONResponseSerializer.init(forInstancesOf: WMFAuthManagerInfo2.self, fromKeypath: "query.authmanagerinfo")
        
        let params = [
            "action": "query",
            "meta": "authmanagerinfo",
            "amirequestsfor": type,
            "format": "json"
        ]
        manager.post("/w/api.php", parameters: params, progress: nil, success: completion, failure: failure)
    }
}
